### DISCUSSION TITLE
Saving intermediate values in a file

### DISCUSSION THREAD

**Discussion by simlaharma** [jax-ml/jax#16990] [Q&A] [answered]
Hello,

I have a JAX function that I use during training a DNN model. I want to debug it by checking the values before and after the function.
So I have been looking for a solution to save the intermediate values during training. The only solution I came across was to use jax.debug.print. However, this only prints and I cannot print huge arrays. Is there a way similar to jax.numpy.save function but for traced arrays?

**Comment by JiaYaobo** [comment, marked as answer]
`jax.experimental.io_callback`?
```python
def save_array(array):
    np.save('array.npy', array)
    print('saved array')

@jax.jit
def f(x):
    io_callback(save_array, None, x)
    return jax.numpy.sin(x)
```

**Reply by jakevdp** [reply]
This is the answer. For more information on callbacks and why they're necessary, see https://jax.readthedocs.io/en/latest/notebooks/external_callbacks.html.
